<a href="https://www.kaggle.com/code/shankubag/placement-predicon?scriptVersionId=240124892" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

<a href="https://colab.research.google.com/github/Shnku/pythoning_stuff/blob/main/mathml/Placement_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🎯Placement Data Prediction: KNN & Logistic Regression


---

# 👨‍💻INTRODUCTION:

This project utilizes the **Placement Prediction Dataset** from Kaggle, which comprises students' academic records and training details.

* [Predicting student placements based on historical data on kaggle](https://www.kaggle.com/datasets/ruchikakumbhar/placement-prediction-dataset)

* The objective is to predict whether a student will be placed based on features such as CGPA, specialization, and training experience.

* Two machine learning techniques used:
  1. K-Nearest Neighbors (KNN)
  2. Logistic Regression

* Goal:  We aim to analyze and compare their effectiveness in forecasting placement outcomes

---


# Importing the Data from kaggle

In [1]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.
import kagglehub
download_path = kagglehub.dataset_download('ruchikakumbhar/placement-prediction-dataset')

print('Data source import complete.\nlocation: ', download_path)


Data source import complete.
location:  /kaggle/input/placement-prediction-dataset


In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
data_path=""
for dirname, _, filenames in os.walk(download_path):
    for filename in filenames:
        data_path=os.path.join(dirname, filename)
        print(data_path)

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All"
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/placement-prediction-dataset/placementdata.csv


# Viewing the Data

In [3]:
dataset = pd.read_csv(data_path)
dataset

StudentID  CGPA  Internships  Projects  Workshops/Certifications  \
0             1   7.5            1         1                         1   
1             2   8.9            0         3                         2   
2             3   7.3            1         2                         2   
3             4   7.5            1         1                         2   
4             5   8.3            1         2                         2   
...         ...   ...          ...       ...                       ...   
9995       9996   7.5            1         1                         2   
9996       9997   7.4            0         1                         0   
9997       9998   8.4            1         3                         0   
9998       9999   8.9            0         3                         2   
9999      10000   8.4            0         1                         1   

      AptitudeTestScore  SoftSkillsRating ExtracurricularActivities  \
0                    65               4.4                        No   
1                    90               4.0                       Yes   
2                    82               4.8                       Yes   
3                    85               4.4                       Yes   
4                    86               4.5                       Yes   
...                 ...               ...                       ...   
9995                 72               3.9                       Yes   
9996                 90               4.8                        No   
9997                 70               4.8                       Yes   
9998                 87               4.8                       Yes   
9999                 66               3.8                        No   

     PlacementTraining  SSC_Marks  HSC_Marks PlacementStatus  
0                   No         61         79       NotPlaced  
1                  Yes         78         82          Placed  
2                   No         79         80       NotPlaced  
3                  Yes         81         80          Placed  
4                  Yes         74         88          Placed  
...                ...        ...        ...             ...  
9995                No         85         66       NotPlaced  
9996                No         84         67          Placed  
9997               Yes         79         81          Placed  
9998               Yes         71         85          Placed  
9999                No         62         66       NotPlaced  

[10000 rows x 12 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   StudentID                  10000 non-null  int64  
 1   CGPA                       10000 non-null  float64
 2   Internships                10000 non-null  int64  
 3   Projects                   10000 non-null  int64  
 4   Workshops/Certifications   10000 non-null  int64  
 5   AptitudeTestScore          10000 non-null  int64  
 6   SoftSkillsRating           10000 non-null  float64
 7   ExtracurricularActivities  10000 non-null  object 
 8   PlacementTraining          10000 non-null  object 
 9   SSC_Marks                  10000 non-null  int64  
 10  HSC_Marks                  10000 non-null  int64  
 11  PlacementStatus            10000 non-null  object 
dtypes: float64(2), int64(7), object(3)
memory usage: 937.6+ KB


# Now we need to identify features(X value) and target(y value)

**Here are 12 columns**

**The last column `PlacementStatus` is `Y` or `target`**

**Rest columns `0-10` are `feature` or `x` values**

In [5]:
x = dataset.drop('PlacementStatus', axis=1) #delete the last (11 th) column
y = dataset['PlacementStatus'] #last column is y value or target

print("x or fearures are:\n ",x.columns ,'\n')
print("y or target is:" ,y.name)
print(y)

x or fearures are:
  Index(['StudentID', 'CGPA', 'Internships', 'Projects',
       'Workshops/Certifications', 'AptitudeTestScore', 'SoftSkillsRating',
       'ExtracurricularActivities', 'PlacementTraining', 'SSC_Marks',
       'HSC_Marks'],
      dtype='object') 

y or target is: PlacementStatus
0       NotPlaced
1          Placed
2       NotPlaced
3          Placed
4          Placed
          ...    
9995    NotPlaced
9996       Placed
9997       Placed
9998       Placed
9999    NotPlaced
Name: PlacementStatus, Length: 10000, dtype: object


## here the `y` data in `Placed` `NotPlaced` not in  0,1 value

**so we need to change it** by pd `replace` function


In [6]:
y_refined = y.replace(['Placed','NotPlaced'], [1,0]).astype(int) #later use map insted
y_refined

/tmp/ipykernel_14/3487602336.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  y_refined = y.replace(['Placed','NotPlaced'], [1,0]).astype(int) #later use map insted


0       0
1       1
2       0
3       1
4       1
       ..
9995    0
9996    1
9997    1
9998    1
9999    0
Name: PlacementStatus, Length: 10000, dtype: int64

In [7]:
y_refined.info()

<class 'pandas.core.series.Series'>
RangeIndex: 10000 entries, 0 to 9999
Series name: PlacementStatus
Non-Null Count  Dtype
--------------  -----
10000 non-null  int64
dtypes: int64(1)
memory usage: 78.3 KB


## Just found also 2 features values are in string form `ExtracurricularActivities` and `PlacementTraining`. 😓

In [8]:
# x_refined = x.replace(['Yes','No'], [1,0]).astype(int)
#! .astype(int) converts all the values ...

x_refined=x
# display(x_refined.head())
x_refined["ExtracurricularActivities"] = x["ExtracurricularActivities"].map({'Yes':1, 'No':0})
x_refined["PlacementTraining"] = x["PlacementTraining"].map({'Yes':1, 'No':0})

x_refined.head()

StudentID  CGPA  Internships  Projects  Workshops/Certifications  \
0          1   7.5            1         1                         1   
1          2   8.9            0         3                         2   
2          3   7.3            1         2                         2   
3          4   7.5            1         1                         2   
4          5   8.3            1         2                         2   

   AptitudeTestScore  SoftSkillsRating  ExtracurricularActivities  \
0                 65               4.4                          0   
1                 90               4.0                          1   
2                 82               4.8                          1   
3                 85               4.4                          1   
4                 86               4.5                          1   

   PlacementTraining  SSC_Marks  HSC_Marks  
0                  0         61         79  
1                  1         78         82  
2                  0         79         80  
3                  1         81         80  
4                  1         74         88

In [9]:
x_refined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   StudentID                  10000 non-null  int64  
 1   CGPA                       10000 non-null  float64
 2   Internships                10000 non-null  int64  
 3   Projects                   10000 non-null  int64  
 4   Workshops/Certifications   10000 non-null  int64  
 5   AptitudeTestScore          10000 non-null  int64  
 6   SoftSkillsRating           10000 non-null  float64
 7   ExtracurricularActivities  10000 non-null  int64  
 8   PlacementTraining          10000 non-null  int64  
 9   SSC_Marks                  10000 non-null  int64  
 10  HSC_Marks                  10000 non-null  int64  
dtypes: float64(2), int64(9)
memory usage: 859.5 KB


# Methods

## 📊K-Nearest Neighbors (KNN)

- Non-parametric algorithm
- Finds closest similar instances in feature space
- Simple implementation, effective for small datasets
- Pros:
  - Intuitive understanding
  - Handles non-linear relationships well
- Cons:
  - Computationally expensive for large datasets
  - Sensitive to noise in data


## 📊Logistic Regression

- Parametric algorithm
- Models probability of binary outcomes
- Linear model with sigmoid activation function
- Pros:
  - Fast computation
  - Interpretable coefficients
  - Works well with high-dimensional data
- Cons:
  - Assumes linear relationship between features and outcome
  - May overfit if not regularized


---

## 📈 Model Evaluation Metrics

* **Confusion Matrix**
  A table showing:

  * True Positives (TP)
  * True Negatives (TN)
  * False Positives (FP)
  * False Negatives (FN)

* **Accuracy**
  Proportion of correct predictions:

  $$
  \text{Accuracy} = \frac{TP + TN}{TP + TN + FP + FN}
  $$

* **Precision**
  Correct positive predictions out of all predicted positives:

  $$
  \text{Precision} = \frac{TP}{TP + FP}
  $$

---


# Main part (Importing Libraries)

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

# Split the dataset to Train and Test

In [11]:
train_x,test_x, train_y, test_y = train_test_split(x_refined,y_refined, test_size=0.2, random_state=42)

In [12]:
from IPython.display import display
display(train_x.head())
display(train_y.head())
display(test_x.head())
display(test_y.head())

StudentID  CGPA  Internships  Projects  Workshops/Certifications  \
9254       9255   8.0            1         3                         3   
1561       1562   7.6            1         2                         1   
1670       1671   8.2            1         3                         2   
6087       6088   7.7            1         1                         0   
6669       6670   8.5            2         3                         2   

      AptitudeTestScore  SoftSkillsRating  ExtracurricularActivities  \
9254                 87               4.6                          1   
1561                 87               3.8                          1   
1670                 86               4.4                          1   
6087                 64               4.1                          0   
6669                 90               4.8                          1   

      PlacementTraining  SSC_Marks  HSC_Marks  
9254                  0         83         88  
1561                  1         61         85  
1670                  1         76         65  
6087                  0         55         59  
6669                  0         80         87

9254    1
1561    0
1670    1
6087    0
6669    1
Name: PlacementStatus, dtype: int64

StudentID  CGPA  Internships  Projects  Workshops/Certifications  \
6252       6253   7.6            0         1                         0   
4684       4685   7.5            1         1                         0   
1731       1732   6.6            1         1                         1   
4742       4743   8.5            1         3                         2   
4521       4522   8.2            1         0                         2   

      AptitudeTestScore  SoftSkillsRating  ExtracurricularActivities  \
6252                 80               3.9                          0   
4684                 81               3.8                          1   
1731                 60               4.0                          0   
4742                 90               4.8                          1   
4521                 90               4.7                          1   

      PlacementTraining  SSC_Marks  HSC_Marks  
6252                  0         60         71  
4684                  1         56         58  
1731                  0         55         64  
4742                  1         88         87  
4521                  1         73         88

6252    0
4684    0
1731    0
4742    0
4521    1
Name: PlacementStatus, dtype: int64

# Results

## Using Logistic Regression Classifier

In [13]:
classifier = LogisticRegression(max_iter=110) 
print(classifier.get_params()) #getting max iter issue for default 100 

{'C': 1.0, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': None, 'max_iter': 110, 'multi_class': 'auto', 'n_jobs': None, 'penalty': 'l2', 'random_state': None, 'solver': 'lbfgs', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}


In [14]:
## Training the model (using `fit`)
classifier.fit(train_x,train_y)

LogisticRegression(max_iter=110)

In [15]:
y_pred_logi = classifier.predict(test_x)
print(y_pred_logi)

[0 0 0 ... 1 1 0]


In [16]:
logi_accuracy = accuracy_score(test_y, y_pred_logi)
print("logistic regression Accuracy:", logi_accuracy)
print(classification_report(test_y,y_pred_logi))

logistic regression Accuracy: 0.7675
              precision    recall  f1-score   support

           0       0.81      0.79      0.80      1172
           1       0.71      0.74      0.72       828

    accuracy                           0.77      2000
   macro avg       0.76      0.76      0.76      2000
weighted avg       0.77      0.77      0.77      2000



## Using KNN

In [17]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(train_x, train_y)

KNeighborsClassifier()

In [18]:
y_pred_knn = knn.predict(test_x)
print(y_pred_knn)

[0 1 0 ... 1 0 0]


In [19]:
knn_accuracy = accuracy_score(test_y, y_pred_knn)
print("KNN Accuracy:", knn_accuracy)
print(classification_report(test_y,y_pred_knn))

KNN Accuracy: 0.7405
              precision    recall  f1-score   support

           0       0.78      0.77      0.78      1172
           1       0.68      0.70      0.69       828

    accuracy                           0.74      2000
   macro avg       0.73      0.73      0.73      2000
weighted avg       0.74      0.74      0.74      2000



## Using Normalized KNN

In [20]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x_train_scaled = scaler.fit_transform(train_x)
x_test_scaled = scaler.fit_transform(test_x)
# Step 4: Fit the KNN model
knn2 = KNeighborsClassifier()
knn2.fit(x_train_scaled, train_y)

# Step 5: Make predictions
y_pred_norm = knn2.predict(x_test_scaled)

# Step 6: Evaluate the model
knn_norm = accuracy_score(test_y, y_pred_norm)
print('normalized accuracy= ',knn_norm)
print(classification_report(test_y,y_pred_norm))

normalized accuracy=  0.772
              precision    recall  f1-score   support

           0       0.80      0.81      0.81      1172
           1       0.73      0.72      0.72       828

    accuracy                           0.77      2000
   macro avg       0.77      0.76      0.76      2000
weighted avg       0.77      0.77      0.77      2000





---
# ✅ Conclusion
Here's a concise conclusion slide for your presentation:

* Logistic Regression achieved an accuracy of **76.8%**
* K-Nearest Neighbors (KNN) achieved **73.9%**
* KNN with feature scaling (Normalization) improved to **76.85%**

**Accuracy changed after fixing the all `.astype(int)` conversion and letting them to be `floar64` and**
* Logistic Regression achieved an accuracy of **76.75%**
* K-Nearest Neighbors (KNN) achieved **74.05%**
* KNN with feature scaling (Normalization) improved to **77.2%**

📌 While both models performed similarly after scaling,
feature normalization significantly improved KNN’s performance,
bringing it on par with Logistic Regression.

🔍 Feature scaling is essential for distance-based models like KNN.


